### DISCUSSION TITLE
Show & Tell 48.3: ExperienceReplayBuffer — Episodic Memory & Gradient Constraints

### DISCUSSION THREAD

**Discussion by web3guru888** [web3guru888/asi-build#947] [Show and tell]
### Why Replay Works

Replay methods address catastrophic forgetting by interleaving stored (or generated) samples from previous tasks during new task training. This implicitly approximates joint training on all tasks, which is the gold standard but requires retaining all data.

### Replay Strategy Comparison

| Method | What is Stored | Memory | Forgetting | Class-IL |
|--------|---------------|--------|------------|----------|
| **Episodic Replay** | Raw inputs + labels | O(K×T) | Low | Good |
| **Generative Replay** | Generator model | Constant | Medium | Good |
| **GEM** | Exemplars + gradients | O(K×T) | Low | Fair |
| **A-GEM** | Exemplar avg gradient | O(K×T) | Medium | Fair |
| **Dark Replay** | Inputs + logits | O(K×T×C) | Very Low | Excellent |

### GEM Gradient Projection Explained

GEM ensures that gradient updates for the current task do not increase loss on previous tasks:

```
minimize   ||g̃ - g||²
subject to g̃ᵀ · g_k ≥ 0,  ∀k ∈ previous tasks
```

This is a quadratic program solved in the dual with T-1 variables (one per previous task). A-GEM approximates this by projecting against the average reference gradient.

### iCaRL Herding: Smart Exemplar Selection

Rather than random selection, iCaRL uses **herding** to select exemplars that best approximate the class mean in feature space. This gives better coverage than random or entropy-based selection with the same memory budget.

### Dark Experience Replay: Storing Knowledge, Not Just Data

DER stores model logits alongside inputs, enabling knowledge distillation during replay:

```
L = CE(f(x_new), y_new) + α · MSE(f(x_replay), z_stored)
```

This preserves not just the correct answer but the full predictive distribution — capturing inter-class relationships that CE loss alone would lose.

Share your replay implementations and buffer management strategies below!

Related: Issue #940 | Planning: #937
